### PR TITLE
Add `distance` member function to nearest predicate

### PR DESCRIPTION
**Pull request by dalg24**
Drafted what we discussed in https://github.com/arborx/ArborX/pull/675#issuecomment-1191882034 so we can think where we want the customization point

### CODE REVIEW

**Comment by aprokop** [comment]
If you want to reconcile this with #675, I assume you would add `distance(int)` member function, as you don't really have access to the geometry? And if we implement APIv2, you add `distance(value)`?

**Review by masterleinad — COMMENTED** [review]
So the idea is for the user to derive from the existing predicates and hide the distance function in the base class by a custom overload?

File: src/details/ArborX_Predicates.hpp [hunk 1]
```diff
@@ -49,12 +49,19 @@ struct Nearest
...
       : _geometry(geometry)
       , _k(k)
   {}
 
+  template <class Other>
+  KOKKOS_FUNCTION auto distance(Other const &other) const
+  {
+    using Details::distance;
+    return distance(_geometry, other);
+  }
```
Comment (aprokop) [inline src/details/ArborX_Predicates.hpp:63]:
```suggestion
  template <class OtherGeometry>
  KOKKOS_FUNCTION auto distance(OtherGeometry const &geometry) const
  {
    using Details::distance;
    return distance(_geometry, geometry);
  }
```

File: src/details/ArborX_Predicates.hpp [hunk 2]
```diff
@@ -90,11 +97,18 @@ struct OrderedSpatial
...
   OrderedSpatial(Geometry const &geometry)
       : _geometry(geometry)
   {}
 
+  template <class Other>
+  KOKKOS_FUNCTION auto distance(Other const &other) const
+  {
+    using Details::distance;
+    return distance(_geometry, other);
+  }
```
Comment (aprokop) [inline src/details/ArborX_Predicates.hpp:110]: Same comment as above.

**Review by aprokop — COMMENTED** [review]
Overall looks good. I think it would be cleaner to rename the template arguments to include `Geometry` in the name.

**Review by aprokop — APPROVED** [review]

File: src/details/ArborX_Predicates.hpp (hunk 1 above)
Comment (dalg24) [reply src/details/ArborX_Predicates.hpp:63]: Can do.  I was going for consistency with https://github.com/arborx/ArborX/blob/843553c9d6ba62700bc9d711a7aec0e0631fea77/src/details/ArborX_Predicates.hpp#L73-L78

Want me to update it all?

Comment (aprokop) [reply src/details/ArborX_Predicates.hpp:63]: Please.

Comment (dalg24) [reply src/details/ArborX_Predicates.hpp]: Done

**Pull request merged**